### PR TITLE
AEA-516 Validating protocol specification

### PR DESCRIPTION
**Pull request by 5A11**
This PR significantly improves the validation of protocol specification before generation. 
Also includes further separation of concern amongst generator modules.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/test_protocols/test_generator.py [hunk 1]
```diff
@@ -468,19 +467,19 @@ def setUp(self):
...
-    def test__includes_custom_type_positive(self, *mocks):
-        """Test _includes_custom_type method positive result."""
-        content_type = "pt:union[pt:str]"
-        result = not _is_composition_type_with_custom_type(content_type)
-        self.assertTrue(result)
-
-        content_type = "pt:optional[pt:str]"
-        result = not _is_composition_type_with_custom_type(content_type)
-        self.assertTrue(result)
+    # @mock.patch(
```
Comment (DavidMinarsch) [inline tests/test_protocols/test_generator.py:470]: Why removed?

File: aea/protocols/generator/validate.py [hunk 2]
```diff
@@ -18,60 +18,183 @@
...
+        or _is_valid_pt(content_type)
+        or _is_valid_set(content_type)
+        or _is_valid_list(content_type)
+        or _is_valid_dict(content_type)
+        or _is_valid_union(content_type)
+        or _is_valid_optional(content_type)
+    )
 
 
 def validate(protocol_specification: ProtocolSpecification) -> Tuple[bool, str]:
```
Comment (DavidMinarsch) [inline aea/protocols/generator/validate.py:200]: This function could be further broken down (essentially along each comment block)

File: tests/test_protocols/test_generator.py (hunk 1 above)
Comment (5A11) [reply tests/test_protocols/test_generator.py:470]: This one tested `_is_composition_type_with_custom_type` which was removed in favour of the more comprehensive validation function.

File: aea/protocols/generator/validate.py (hunk 2 above)
Comment (5A11) [reply aea/protocols/generator/validate.py]: Done

File: aea/protocols/generator/common.py [hunk 3]
```diff
@@ -214,8 +332,8 @@ def try_run_protoc(path_to_generated_protocol_package, name) -> None:
             "--python_out={}".format(path_to_generated_protocol_package),
             "{}/{}.proto".format(path_to_generated_protocol_package, name),
         ],
-        capture_output=True,
-        text=True,
+        stderr=subprocess.PIPE,
```
Comment (5A11) [inline aea/protocols/generator/common.py:335]: Addressing python 3.6 compatibility issue.

**Pull request merged**